scodec.stream.CodecError: unexpected end of input

## What I did

Extract [scalarmult8.zip](https://github.com/satabin/swam/files/4911244/scalarmult8.zip) to the cwd and execute `./millw cli.run run --wasi --dir=. scalarmult8`

Also, I verified that wasmer can execute the file

## What I get

```
swam.runtime.CompileException: An error occurred during compilation
	at swam.runtime.internals.compiler.Compiler.$anonfun$compile$15(Compiler.scala:236)
	at swam.runtime.internals.compiler.Compiler.$anonfun$compile$15$adapted(Compiler.scala:236)
	at fs2.Stream$.$anonfun$handleErrorWith$1(Stream.scala:1551)
	at fs2.internal.FreeC$$anon$5.cont(FreeC.scala:68)
	at fs2.internal.FreeC$ViewL$$anon$9$$anon$10.cont(FreeC.scala:203)
	at fs2.internal.FreeC$ViewL$.mk(FreeC.scala:192)
	at fs2.internal.FreeC$ViewL$.apply(FreeC.scala:183)
	at fs2.internal.FreeC.viewL(FreeC.scala:83)
	at fs2.internal.Algebra$.go$1(Algebra.scala:178)
	at fs2.internal.Algebra$.$anonfun$compileLoop$16(Algebra.scala:310)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:145)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:366)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:387)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:330)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:141)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:366)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:387)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:330)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at cats.effect.internals.PoolUtils$$anon$2$$anon$3.run(PoolUtils.scala:52)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: scodec.stream.CodecError: unexpected end of input
```

PS:
- Once again from the [libsodium testsuite](https://github.com/jedisct1/libsodium/)
- Binaries not stripped this time
- All other tests pass :100:

--- a/cli/src/swam/cli/Main.scala
+++ b/cli/src/swam/cli/Main.scala
@@ -54,6 +54,9 @@ object Main extends CommandIOApp(name = "swam-cli", header = "Swam from the comm
   val wasmFile =
     Opts.argument[Path](metavar = "wasm")
 
+  val readChunkSize =
+    Opts.option[Int]("read-chunk-size", "Size in bytes of the WASM file read buffer (default 1024)").withDefault(1024)
+
   val restArguments =
     Opts.arguments[String](metavar = "args").orEmpty
 
@@ -115,18 +118,30 @@ object Main extends CommandIOApp(name = "swam-cli", header = "Swam from the comm
     .option[Path]("out", "Save decompiled result in the given file. Prints to stdout if not provider", short = "o")
 
   val runOpts: Opts[Options] = Opts.subcommand("run", "Run a WebAssembly file") {
-    (mainFun, wat, wasi, wasiOptions, time, dirs, trace, traceFile, filter, debug, wasmFile, restArguments).mapN {
-      (main, wat, wasi, wasiOptions, time, dirs, trace, traceFile, filter, debug, wasm, args) =>
-        Run(wasm, args, main, wat, wasi, wasiOptions, time, trace, filter, traceFile, dirs, debug)
+    (mainFun,
+     wat,
+     wasi,
+     wasiOptions,
+     time,
+     dirs,
+     trace,
+     traceFile,
+     filter,
+     debug,
+     wasmFile,
+     readChunkSize,
+     restArguments).mapN {
+      (main, wat, wasi, wasiOptions, time, dirs, trace, traceFile, filter, debug, wasm, readChunkSize, args) =>
+        Run(wasm, args, main, wat, wasi, wasiOptions, time, trace, filter, traceFile, dirs, debug, readChunkSize)
     }
   }
 
   val decompileOpts: Opts[Options] = Opts.subcommand("decompile", "Decompile a wasm file") {
-    (textual, wasmFile, out.orNone).mapN { (textual, wasm, out) => Decompile(wasm, textual, out) }
+    (wasmFile, textual, out.orNone, readChunkSize).mapN(Decompile(_, _, _, _))
   }
 
   val validateOpts: Opts[Options] = Opts.subcommand("validate", "Validate a wasm file") {
-    (wasmFile, wat, dev).mapN(Validate(_, _, _))
+    (wasmFile, wat, dev, readChunkSize).mapN(Validate(_, _, _, _))
   }
 
   val compileOpts: Opts[Options] = Opts.subcommand("compile", "Compile a wat file to wasm") {
@@ -174,7 +189,7 @@ object Main extends CommandIOApp(name = "swam-cli", header = "Swam from the comm
     runOpts.orElse(decompileOpts).orElse(validateOpts).orElse(compileOpts).map { opts =>
       Blocker[IO].use { blocker =>
         opts match {
-          case Run(file, args, main, wat, wasi, wasiOptions, time, trace, filter, tracef, dirs, debug) =>
+          case Run(file, args, main, wat, wasi, wasiOptions, time, trace, filter, tracef, dirs, debug, readChunkSize) =>
             for {
               tracer <- if (trace)
                 JULTracer[IO](blocker,
@@ -186,17 +201,18 @@ object Main extends CommandIOApp(name = "swam-cli", header = "Swam from the comm
                 IO(None)
               engine <- Engine[IO](blocker, tracer)
               tcompiler <- swam.text.Compiler[IO](blocker)
-              module = if (wat) tcompiler.stream(file, debug, blocker) else engine.sections(file, blocker)
+              module = if (wat) tcompiler.stream(file, debug, blocker)
+              else engine.sections(file, blocker, chunkSize = readChunkSize)
               compiled <- engine.compile(module)
               _ <- doRun(compiled, main, dirs, args, wasi, wasiOptions, time, blocker)
             } yield ExitCode.Success
-          case Decompile(file, textual, out) =>
+          case Decompile(file, textual, out, readChunkSize) =>
             for {
               decompiler <- if (textual)
                 TextDecompiler[IO](blocker)
               else
                 RawDecompiler[IO]
-              doc <- decompiler.decompilePath(file, blocker)
+              doc <- decompiler.decompilePath(file, blocker, chunkSize = readChunkSize)
               outPipe = out.fold(fs2.io.stdout[IO](blocker))(fs2.io.file.writeAll[IO](_, blocker, outFileOptions))
               _ <- Stream
                 .emits(doc.render(10).getBytes("utf-8"))
@@ -204,7 +220,7 @@ object Main extends CommandIOApp(name = "swam-cli", header = "Swam from the comm
                 .compile
                 .drain
             } yield ExitCode.Success
-          case Validate(file, wat, dev) =>
+          case Validate(file, wat, dev, readChunkSize) =>
             val throwableFormat =
               if (dev)
                 ThrowableFormat(ThrowableFormat.Depth.Full, ThrowableFormat.Indent.Fixed(2))
@@ -215,7 +231,8 @@ object Main extends CommandIOApp(name = "swam-cli", header = "Swam from the comm
             for {
               engine <- Engine[IO](blocker)
               tcompiler <- swam.text.Compiler[IO](blocker)
-              module = if (wat) tcompiler.stream(file, false, blocker) else engine.sections(file, blocker)
+              module = if (wat) tcompiler.stream(file, false, blocker)
+              else engine.sections(file, blocker, chunkSize = readChunkSize)
               res <- engine.validate(module).attempt
               _ <- res.fold(t => logger.error("Module is invalid", t), _ => logger.info("Module is valid"))
             } yield ExitCode.Success

--- a/cli/src/swam/cli/options.scala
+++ b/cli/src/swam/cli/options.scala
@@ -34,11 +34,12 @@ case class Run(file: Path,
                traceFilter: String,
                traceFile: Path,
                dirs: List[Path],
-               debug: Boolean)
+               debug: Boolean,
+               readChunkSize: Int)
     extends Options
 
-case class Decompile(file: Path, textual: Boolean, out: Option[Path]) extends Options
+case class Decompile(file: Path, textual: Boolean, out: Option[Path], readChunkSize: Int) extends Options
 
-case class Validate(file: Path, wat: Boolean, dev: Boolean) extends Options
+case class Validate(file: Path, wat: Boolean, dev: Boolean, readChunkSize: Int) extends Options
 
 case class Compile(file: Path, out: Path, debug: Boolean) extends Options

--- a/core/src/swam/binary/InstCodec.scala
+++ b/core/src/swam/binary/InstCodec.scala
@@ -487,11 +487,11 @@ trait InstCodec extends TypeCodec {
         case MemorySize =>
           constant(hex"00")
             .encode(())
-            .map(_ => BitVector.fromByte(OpCode.MemorySize.toByte))
+            .map(BitVector.fromByte(OpCode.MemorySize.toByte) ++ _)
         case MemoryGrow =>
           constant(hex"00")
             .encode(())
-            .map(_ => BitVector.fromByte(OpCode.MemoryGrow.toByte))
+            .map(BitVector.fromByte(OpCode.MemoryGrow.toByte) ++ _)
         case i32.Const(v) =>
           varint32
             .encode(v)

--- a/core/src/swam/binary/WasmCodec.scala
+++ b/core/src/swam/binary/WasmCodec.scala
@@ -110,8 +110,8 @@ object WasmCodec extends InstCodec {
 
   val custom: Codec[(String, BitVector)] =
     variableSizeBytes(varuint32,
-                      (("name" | variableSizeBytes(varuint32, utf8)) ~
-                        ("payload" | scodec.codecs.bits)))
+                      ("name" | variableSizeBytes(varuint32, utf8)) ~
+                        ("payload" | scodec.codecs.bits))
 
   val section =
     discriminated[Section]

--- a/core/src/swam/binary/varint.scala
+++ b/core/src/swam/binary/varint.scala
@@ -46,7 +46,7 @@ private class Varuint(bits: Int) extends Codec[Int] {
     if (n <= 0) {
       Attempt.failure(Err("integer representation too long"))
     } else if (buffer.isEmpty) {
-      Attempt.failure(Err("unexpected end of input"))
+      Attempt.failure(Err.insufficientBits(8, 0))
     } else {
       val byte = buffer.head
       if (n >= 7 || (byte & 0x7f) < (1 << n)) {
@@ -95,8 +95,8 @@ private class Varint(bits: Int) extends Codec[Long] {
 
   private val bitsL = bits.toLong
 
-  val MaxValue = (1 << (bits - 1)) - 1
-  val MinValue = -(1 << (bits - 1))
+  val MaxValue = if (bits == 64) Long.MaxValue else (1L << (bits - 1)) - 1
+  val MinValue = if (bits == 64) Long.MinValue else -(1L << (bits - 1))
 
   private def description = s"$bits-bit signed leb128 integer"
 
@@ -110,7 +110,7 @@ private class Varint(bits: Int) extends Codec[Long] {
     if (n <= 0) {
       Attempt.failure(Err("integer representation too long"))
     } else if (buffer.isEmpty) {
-      Attempt.failure(Err("unexpected end of input"))
+      Attempt.failure(Err.insufficientBits(8, 0))
     } else {
       val byte = buffer.head
       val mask = (-1 << (n - 1)) & 0x7f

